File: 142. Linked List Cycle II.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
	set := map[*ListNode]bool{}
	cur := head
	for cur != nil {
		if _, didFind := set[cur]; didFind {
			return cur
		}
		set[cur] = true
		cur = cur.Next
	}
	return nil
}

File: 142. Linked List Cycle II.py
from typing import List, Optional

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:

        slow = head
        fast = head

        intersection = None

        while slow != None and fast != None and fast.next != None:

            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                intersection = slow
                break

        if intersection == None:
            return None

        slow = head

        while intersection != slow:
            slow = slow.next
            intersection = intersection.next

        return intersection

